pkg/controllers/provisioning/v1alpha1/reallocator/controller.go

```diff
@@ -53,31 +54,60 @@ func (c *Controller) Name() string {
 func NewController(kubeClient client.Client, cloudProvider cloudprovider.Factory) *Controller {
 	return &Controller{
 		filter:        &Filter{kubeClient: kubeClient},
+		terminator:    &Terminator{kubeClient: kubeClient},
 		cloudProvider: cloudProvider,
 	}
 }
 
 // Reconcile executes an allocation control loop for the resource
 func (c *Controller) Reconcile(object controllers.Object) error {
+	provisioner := object.(*v1alpha1.Provisioner)
 	ctx := context.TODO()
-	// 1. Filter all nodes with Karpenter TTL label
-	// TODO: Filter only nodes with the Provisioner label
 
-	// 2. Remove TTL from nodes with TTL label that have pods and if past TTL, cordon/drain node
-
-	// 3. Filter under-utilized nodes
-	nodes, err := c.filter.GetUnderutilizedNodes(ctx)
+	// 1. Get underutilized nodes
+	underutilized, err := c.filter.GetUnderutilizedNodes(ctx, provisioner)
 	if err != nil {
-		return fmt.Errorf("filtering nodes, %w", err)
+		return fmt.Errorf("listing underutilized nodes, %w", err)
+	}
+
+	// TODO: Further filter underutilized nodes that haven't been cordoned/TTLed to not spam logs
+	if len(underutilized) != 0 {
+		zap.S().Debugf("Found %d underutilized nodes", len(underutilized))
+	}
+
+	// 2. Set TTL on underutilized nodes
+	// TODO: Go routines to parllelize AddTTL
+	for _, node := range underutilized {
+		if err := c.terminator.AddTTL(ctx, node); err != nil {
+			return fmt.Errorf("adding ttl, %w", err)
+		}
 	}
-	if len(nodes) == 0 {
-		return nil
+
+	// 3. Find Nodes Past TTL with Karpenter Labels
+	expiredNodes, err := c.filter.GetExpiredNodes(ctx, provisioner)
+	if err != nil {
+		return fmt.Errorf("getting TTLed nodes, %w", err)
 	}
-	zap.S().Infof("Found %d underutilized nodes", len(nodes))
 
 	// 4. Cordon each node
+	// TODO: Go routines to parallelize CordonNode
```

Comment:
You could imagine cordon takes a []Node which would allow you to hide the parallelization details inside the Cordon function.

Reply:
Yeah. That would be a good idea too, could do that for the TTL, and deleting as well!